tests/e2e/macvtap_test.go

```diff
@@ -0,0 +1,290 @@
+
+				})
+
+				Context("WHEN another VMI is created, having an additional macvtap interface", func() {
+					It("THEN it can reach another VMI via the macvtap interface", func() {
+
+					})
+
+					It("THEN it cannot reach the host where its pod was scheduled", func() {
+
+					})
+
+					It("THEN it can reach the other hosts in the cluster", func() {
+
+					})
+				})
+			})
+		})
+	})
+})
+
+func deleteNetworkAttachmentDefinition(macvtapIfaceName string, namespace string) rest.Result {
+	return clientset.RESTClient().
+		Delete().
+		RequestURI(fmt.Sprintf(postUrl, namespace, macvtapIfaceName)).
+		Do()
+}
+
+func provisionNetworkAttachmentDefinition(macvtapIfaceName string, lowerDeviceName string, namespace string) rest.Result {
+	return clientset.RESTClient().
+		Post().
+		RequestURI(fmt.Sprintf(postUrl, namespace, macvtapIfaceName)).
+		Body([]byte(fmt.Sprintf(nad, macvtapIfaceName, namespace, buildMacvtapResourceName(lowerDeviceName), macvtapIfaceName))).
+		Do()
+}
+
+func filterPods(pods []v1.Pod, filterFunction func(v1.Pod) bool) []v1.Pod {
+	filteredPods := make([]v1.Pod, 0)
+	for _, pod := range pods {
+		if filterFunction(pod) {
+			filteredPods = append(filteredPods, pod)
+		}
+	}
+	return filteredPods
+}
+
+func buildMacvtapResourceName(macvtapIfaceName string) string {
+	return fmt.Sprintf("%s/%s", macvtapResource, macvtapIfaceName)
+}
+
+func waitForResourceAvailability(timeout time.Duration, resourceName string) {
+	checkForResourceAvailable := func() bool {
+		nodeList, err := clientset.CoreV1().Nodes().List(metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		nodesWithResourceAvailable := 0
+		for _, node := range nodeList.Items {
+			if _, ok := node.Status.Capacity[v1.ResourceName(resourceName)]; ok {
+				nodesWithResourceAvailable += 1
+			}
+		}
+
+		return nodesWithResourceAvailable == len(nodeList.Items)
+	}
+
+	Eventually(checkForResourceAvailable, timeout, 2*time.Second).Should(BeTrue())
+}
+
+func buildMacvtapNetworkAnnotations(networkName string) map[string]string {
+	requestMacvtapNetwork := make(map[string]string)
+	requestMacvtapNetwork[networkResource] = networkName
+	return requestMacvtapNetwork
+}
+
+func buildMacvtapResourceRequest(resourceName string, quantity int) v1.ResourceList {
+	return v1.ResourceList{
+		v1.ResourceName(buildMacvtapResourceName(resourceName)): resource.MustParse(strconv.Itoa(quantity)),
+	}
+}
+
+// Parse the json network reported by Multus in the networks-status annotations
+func parseNetwork(network string) ([]reportedNetwork, error) {
+	var reportedNetwork []reportedNetwork
+	err := json.Unmarshal([]byte(network), &reportedNetwork)
+	return reportedNetwork, err
+}
+
+func waitForPodReadiness(podName string, namespace string, timeout time.Duration) {
+	isPodReady := func() bool {
+		pod, err := clientset.CoreV1().Pods(namespace).Get(podName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
```

Comment:
It is better to explain what went wrong. `NotTo(HaveOccured(), "Is able to list pods"`

Then we get nicer explanation when this fails. Ginkgo can be really annoying when it fails in a helper function in a body of a test and you have no clue what went wrong. With error message you can at least guess, but with bool assertions (as you do in other functions) you would not learn much from the error output.

Reply:
Done.